zorch/lnp/exact.py

```
@@ -0,0 +1,229 @@
+        anywhere — it is a proof of a statement about residues that reads
+        like a proof about integers."""
+        modulus = math.prod(self.ring.q_moduli)
+        span = math.isqrt(_DIGIT_COLS * self.ring.d)
+        for name, value in (
+            ("the binarity check", projection_bound**2 + span * projection_bound),
+            ("the norm identity", 2 * self.bound**2 + projection_bound**2 - 1),
+        ):
+            if value >= modulus:
+                raise ValueError(
+                    f"exact: {name} needs {value} < q = {modulus}, so the "
+                    f"identity it proves modulo q need not hold over ℤ — "
```

Ruff (0.16.2)

[warning] String contains ambiguous `ℤ` (DOUBLE-STRUCK CAPITAL Z). Did you mean `Z` (LATIN CAPITAL LETTER Z)?

(RUF001)